Consolidate the key schedule (#2)
cfrg/draft-irtf-cfrg-hpke: draft-irtf-cfrg-hpke.md

The procedures described in this session are laid out in a
Python-like pseudocode.  The ciphersuite in use is left implicit.

> ## Key Schedule

Reviewer: I might rename this to Encryption Context, and then the function `KeySchedule` to `EncryptionContext`, respectively. That seems to better match what's happening here.

The variants of HPKE defined in this document share a common

> mechanism for translating the inputs to the protocol into an

Reviewer: ```suggestion
mechanism for translating the protocol inputs into an
```

> encryption context.  The inputs to the key schedule are as follows:

Reviewer: ```suggestion
encryption context.  The key schedule inputs are as follows:
```

* `pkR` - The receiver's public key
* `zz` - A shared secret generated via the KEM for this transaction
* `enc` - An encapsulated key produced by the KEM for the receiver
* `info` (optional) - Application-supplied information
* `psk` (optional) - A pre-shared secret held by both the initiator
  and the receiver.

> * `pskID` (optional) - An identifier for the PSK

Reviewer: Can one supply a `pskID` but not a `psk`?

* `pkI` (optional) - The initiator's public key

> If any of the optional values is not present (e.g., set to `None`

Reviewer: ```suggestion
If any of the optional values is absent (e.g., set to `None`
```

Author: Actually, this whole section needs rewriting after the update.

> in a Python implementation), then it is treated in the same way as

Reviewer: ```suggestion
in a Python implementation), then it is treated as
```

an empty octet string.

The key and nonce produced by this algorithm have the property that

> they are only known to the holder of the receipient private key, and

Reviewer: This seems to suggest that the key and nonce are outputs, but they're just put in the context. Perhaps we replace "produced" with "computed"?

the party that ran the KEM to generate `zz` and `enc`.  If the `psk`
and `pskID` arguments are provided, then the recipient is assured
that the initiator held the PSK.  If the `pkIm` argument is
provided, then the recipient is assued that the initator held the
corresponding private key (assuming that `zz` and `enc` were
generated using the AuthEncap / AuthDecap methods; see below).

~~~~~
default_pkIm = zero(Npk)

> default_salt = zero(Nh)

Reviewer: nit: instead of `default`, perhaps `empty`? And perhaps replace `salt` with `psk`?

Author: Don't like that; "empty" implies there's nothing there, whereas here clearly there is.

default_pskId = zero(0)

def VerifyMode(mode, salt, pskID, pkIm):

>   got_psk = (salt != default_salt and pskID != default_pskID)

Reviewer: nit: can we rename `salt` to `psk`? I had to go down to see its invocation to learn that salt is really the psk here.

Reviewer: (I recognize we use `salt` since it's more natural for `Extract`, though `psk` I think being clear about the nature of the value is probably better here.

  no_psk = (salt == default_salt and pskID == default_pskID)
  got_pkIm = (pkIm != default_pkIm)
  no_pkIm = (pkIm == default_pkIm)

  if mode == mode_base and (got_psk or got_pkIm):
    raise Exception("Invalid configuration for mode_base")
  if mode == mode_psk and (no_psk or got_pkIm):
    raise Exception("Invalid configuration for mode_psk")
  if mode == mode_auth and (got_psk or no_pkIm):
    raise Exception("Invalid configuration for mode_auth")
  if mode == mode_psk_auth and (no_psk or no_pkIm):
    raise Exception("Invalid configuration for mode_psk_auth")

def KeySchedule(mode, pkRm, zz, enc, info, salt, pskID, pkIm):

>   VerifyMode(mode, psk, pskID, pkI)

Reviewer: nit: either this should be `salt`, or the `salt` parameter to `KeySchedule` should be `psk`, right?

  pkRm = Marshal(pkR)
  context = concat(mode, ciphersuite,

>                    len(enc), enc, len(pkRm), pkRm, pkIm,

Reviewer: We don't include `len(pkIm)`?

Author: `pkIm` has fixed length `Npk`.

>                    len(enc), enc, len(pkRm), pkRm, pkIm,

Reviewer: Question: if `Nenc` is fixed for the cipher suite, do we need to encode the length here? (This doesn't seem like it'd match the TLS struct below, since `opaque enc[Nenc];` is a fixed-length array.

Author: Good point.  Removed this length, and the length for `pkRm`, which is unnecessary for similar reasons.
